File: CyFile-Android/app/src/test/java/at/tugraz/tc/cyfile/secret/HashSecretRepositoryUnitTest.java
package at.tugraz.tc.cyfile.secret;

import android.content.Context;

import junit.framework.Assert;

import org.junit.Test;
import org.mockito.Mockito;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.OutputStream;

import static org.mockito.Mockito.mock;

public class HashSecretRepositoryUnitTest {

    private ByteArrayOutputStream createOutputStream() throws IOException {

        return new ByteArrayOutputStream(2048);
    }

    private InputStream createInputStream(HashedSecret secret) throws IOException {
        ByteArrayOutputStream bas = new ByteArrayOutputStream(2048);
        ObjectOutputStream oos = new ObjectOutputStream(bas);
        oos.writeObject(secret);

        byte[] buffer = bas.toByteArray();
        return new ByteArrayInputStream(buffer);
    }

    @Test
    public void testReadSecret() throws IOException {
        HashedSecret secret = new HashedSecret(() -> "111222", new ApacheCodecBase64());
        Context mockContext = mock(Context.class);

        HashSecretRepository repository =
                Mockito.spy(new HashSecretRepository(mockContext, ""
                        , mock(CyFileLogger.class), new ApacheCodecBase64()));

        Mockito.doReturn(createInputStream(secret)).when(repository).getInputStream();
        repository.init();
        Assert.assertEquals(repository.getSecret(), secret);
    }

    @Test
    public void testWriteSecret() throws IOException, ClassNotFoundException {
        HashedSecret oldSecret = new HashedSecret(() -> "332211", new ApacheCodecBase64());
        Secret newSecret = new PinPatternSecret("111222");
        Context mockContext = mock(Context.class);

        HashSecretRepository repository =
                Mockito.spy(new HashSecretRepository(mockContext, ""
                        , mock(CyFileLogger.class), new ApacheCodecBase64()));

        Mockito.doReturn(createInputStream(oldSecret)).when(repository).getInputStream();
        ByteArrayOutputStream outputStream = createOutputStream();
        Mockito.doReturn(outputStream).when(repository).getOutputStream();
        repository.init();

        repository.saveSecret(newSecret);
        byte[] buffer = outputStream.toByteArray();
        ByteArrayInputStream is = new ByteArrayInputStream(buffer);
        ObjectInputStream ois = new ObjectInputStream(is);
        HashedSecret actual = (HashedSecret) ois.readObject();
        Assert.assertEquals(actual, newSecret);
    }

    @Test
    public void testGetInputStream() throws Exception {
        HashedSecret secret = new HashedSecret(() -> "111222", new ApacheCodecBase64());
        Context mockContext = mock(Context.class);

        HashSecretRepository repository =
                Mockito.spy(new HashSecretRepository(mockContext, ""
                        , mock(CyFileLogger.class), new ApacheCodecBase64()));
        Mockito.doReturn(createInputStream(secret)).when(repository).getInputStream();
        repository.init();

        repository.getInputStream();
    }

    @Test
    public void testGetOutputStream() throws Exception {
        HashedSecret secret = new HashedSecret(() -> "111222", new ApacheCodecBase64());
        Context mockContext = mock(Context.class);

        HashSecretRepository repository =
                Mockito.spy(new HashSecretRepository(mockContext, ""
                        , mock(CyFileLogger.class), new ApacheCodecBase64()));

        OutputStream out = createOutputStream();
        Mockito.doReturn(createInputStream(secret)).when(repository).getInputStream();
        Mockito.doReturn(out).when(repository).getOutputStream();
        repository.init();

        OutputStream result = repository.getOutputStream();

        Assert.assertEquals(result, out);
    }
}

File: CyFile-Android/app/src/test/java/at/tugraz/tc/cyfile/secret/HashSecretRepositoryUnitTest.kt
package at.tugraz.tc.cyfile.secret

import at.tugraz.tc.cyfile.BaseUnitTest
import at.tugraz.tc.cyfile.crypto.impl.ApacheCodecBase64
import at.tugraz.tc.cyfile.file.FileHandler
import at.tugraz.tc.cyfile.secret.impl.HashSecretRepository
import at.tugraz.tc.cyfile.secret.impl.HashedSecret
import at.tugraz.tc.cyfile.secret.impl.PinPatternSecret
import io.mockk.every
import io.mockk.impl.annotations.RelaxedMockK
import io.mockk.mockk
import junit.framework.Assert
import org.junit.Before
import org.junit.Test
import java.io.*

class HashSecretRepositoryUnitTest : BaseUnitTest() {

    @RelaxedMockK
    private lateinit var fileHandler: FileHandler
    private val defaultSecretValue = "111222"
    private val defaultSecret = HashedSecret(PinPatternSecret(defaultSecretValue), ApacheCodecBase64())

    private lateinit var repository: HashSecretRepository

    @Before
    fun setup() {
        every { fileHandler.getOutputStream() } returns createOutputStream()
        every { fileHandler.getInputStream() } returns createInputStream(
                defaultSecret)
        repository = HashSecretRepository(fileHandler, mockk(relaxed = true), ApacheCodecBase64())
    }

    @Throws(IOException::class)
    private fun createOutputStream(): ByteArrayOutputStream {
        return ByteArrayOutputStream(2048)
    }

    @Throws(IOException::class)
    private fun createInputStream(secret: HashedSecret): InputStream {
        val bas = ByteArrayOutputStream(2048)
        val oos = ObjectOutputStream(bas)
        oos.writeObject(secret)

        val buffer = bas.toByteArray()
        return ByteArrayInputStream(buffer)
    }

    @Test
    @Throws(IOException::class)
    fun testReadSecret() {
        Assert.assertEquals(repository.retrieveSecret(), defaultSecret)
    }

    @Test
    @Throws(IOException::class, ClassNotFoundException::class)
    fun testWriteSecret() {
        val newSecret = PinPatternSecret("332211")

        repository.saveSecret(newSecret)

        val buffer = (fileHandler.getOutputStream() as ByteArrayOutputStream).toByteArray()
        val `is` = ByteArrayInputStream(buffer)
        val ois = ObjectInputStream(`is`)
        val actual = ois.readObject() as HashedSecret
        Assert.assertEquals(actual, newSecret)
    }
}
